### ISSUE TITLE
Calling minetest.emerge_area() on S16 boundary schedules 4096 emerge calls, hangs Minetest server

### ISSUE THREAD

**Issue by erlehmann** [minetest/minetest#12327]
##### Minetest version
At least Minetest 5.5.0 and 5.4.1. Probably all versions of Minetest right now.

##### OS / Hardware
Operating system: GNU/Linux
CPU: both x86 & AMD64

##### Summary
Doing an emerge call with a position where x is 32767 as the first argument hangs the Minetest server by scheduling 4096 emerge calls, using up CPU and RAM resources. I tried to wait until all those emerges were finished, but never had enough patience for it.

##### Steps to reproduce
1. Get the crashtest modpack from here: <git.minetest.land/erlehmann/crashtest>
2. Disable the “emerge_hang_fix” mod.  Make sure the `emerge_hang` mod is enabled.
3. Enter a world.
4. Verify that the world is not loading and you see the number of remaining emerge calls being logged from 4096 downwards.

##### Workaround
1. Install and enable the crashtest modpack from here: <git.minetest.land/erlehmann/crashtest>
2. Lobby ContentDB staff to allow the modpack on CDB, so users can easily install it on existing Minetest versions to avoid the crash.

##### TODO
* Make the `emerge_hang_fix` mod clamp the area to S16 bounds instead of canceling the entire emerge.
* Check with a debugger what actually happens. My analysis above might, after all, be wrong in some parts.

##### TODONT
* If you think that a solution to this issue should involve `MAX_MAP_GENERATION_LIMIT` for bounds checking, consider asking someone for help who understands how to do bounds checking for algorithms taking S16 data types.
* Do not silently avoid emerges with invalid inputs. Logging them hints at errors in mods and at possible attacks by griefers.

**Comment by erlehmann** [comment]
@rubenwardy as this is remotely-triggerable denial of service, I think it would be good to publish an advisory. Your thoughts?

**Comment by Df458** [comment]
Will investigate the code flow later.

**Comment by Df458** [comment]
Yeah, the cause of this is pretty simple. 32768 is read by `read_v3s16` (via `doubleToInt`) as -32768, probably a simple integer overflow. As a result, if you read a number just before the line and just after the line then it will attempt to emerge everything. This happens to be 4096, because dividing by the mapblock size (16) yields 2048 and -2048.

We should probably consider bounds-checking the function, or alternatively bounds-checking the inputs to emerge if `doubleToInt` needs to be as fast as possible...

A few other possible instances of this specific bug:
- `l_find_nodes_in_area` and `l_find_nodes_in_area_under_air`
- `l_get_voxel_manip`
- `l_fix_light`
- `l_delete_area`
- `l_find_path`

**Comment by erlehmann** [comment]
@Df458 oh, this explanation makes sense! While I am a fan of functions checking their inputs and outputs, this seems like a type issue. Do you know if this issue could be fixed all at once in one specific part of the code or does every function have to consider the boundaries separately?

**Comment by sfan5** [comment]
> tried to wait until all those emerges were finished, but never had enough patience for it.

This takes less than 2 minutes on my machine.
